Add stream support to rpc.

sbunce opened the pull request:
Add support for client and server streaming.

Example:
```
service Serv {
	rpc Foo(stream Bar) returns (stream Baz);
}
```

sbunce commented:
David,
My goal is to use gotoc with grpc. Any pointers on what needs to get done are appreciated. Thanks!

Seth

sbunce commented:
I thought of a problem with this CL. There may be a message called "stream". I'll think about how to handle this.
```
message stream {}

service Serv {
    rpc Foo(stream) returns (stream);
}
```

sbunce commented:
I was able to resolve that ambiguity. It now handles the corner case where a message is called "stream".  I added a test for it. The code is more complicated but I can't think of a way to make it simpler.

```
message stream {}

service Foo {
rpc Bar(stream) returns (stream);
}
```

dsymonds commented:
Yeah, this looks too complicated. There's other parts of the parser that effectively do a token lookahead. I think you can just parse the token, and if it is the literal "stream" then the next token is either the stream's message name or ")". Alternatively, read two tokens (there always has to be at least two left), and it'll either be "stream" and Msg, or Msg and ")".

dsymonds commented:
Also, please squash the commits when you push a new version.

sbunce commented:
I checked in some garbage by accident. Trying to fix now. Please ignore last commit.

sbunce commented:
I fixed my bad git push. Sorry for that noise.

I took a look at other parts of the parser where you use p.back(). I refactored this change to use p.back() and a few conditions. I think it's simpler to understand now.

dsymonds commented:
Sorry I've been slow. I'll try to get to it this weekend.

dsymonds commented on parser/parser_test.go:
```diff
@@ -331,7 +331,19 @@ var parseTests = []parseTest{
 		"SimpleService",
 		"service TestService {\n  rpc Foo(In) returns (Out);\n}\n message In{} message Out{}",
-		`service { name: "TestService" method { name:"Foo" input_type:".In" output_type:".Out" } }` +
+		`service { name: "TestService" method { name:"Foo" input_type:".In" output_type:".Out" client_streaming: false server_streaming: false } }` +
```
This shouldn't have changed. client_streaming/server_streaming should not have been set if they were false.

sbunce replied:
Done

dsymonds commented on parser/parser_test.go:
```diff
@@ -331,7 +331,19 @@ var parseTests = []parseTest{
+			`message_type:{name:"In"} message_type:{name:"Out"}`,
+	},
+	{
+		"SimpleServiceWithMessageCalledStream",
```
See src/google/protobuf/compiler/parser_unittest.cc in the protobuf repo for the test cases that I've adapted. I think this just needs MethodsAndStreams ported over.

sbunce replied:
My intention here is to test the special case where there's a message called "stream". I did this because my first version didn't handle this special case and it didn't seem obvious to me. This could just be me though. I took a look at parser_unittest.cc and it doesn't look like it tests for this special case.

dsymonds replied:
Ah, I missed that. Right. Okay, keep this test. You're right that MethodsAndStreams looks pointless; I should have looked more closely.

dsymonds commented on gendesc/gendesc.go:
```diff
@@ -275,9 +275,11 @@ func genService(srv *ast.Service) (*pb.ServiceDescriptorProto, error) {
+		Name:            proto.String(mth.Name),
+		InputType:       proto.String(qualifiedName(mth.InType)),
+		OutputType:      proto.String(qualifiedName(mth.OutType)),
+		ClientStreaming: proto.Bool(mth.ClientStreaming),
```
set these conditionally.

sbunce replied:
Done.

dsymonds commented on parser/parser.go:
```diff
@@ -781,15 +781,29 @@ func (p *parser) readService(srv *ast.Service) *parseError {
 			return tok.err
 		}
-		mth.InTypeName = tok.value // TODO: validate
+		if tok.value == "stream" {
```
What about something more like this?

mth.InTypeName = tok.value // TODO: validate
if tok.value == "stream" {
  // If the next token isn't ")", this is a stream.
  tok = p.next()
  if tok.err != nil { return tok.err }
  if tok.value != ")" {
    mth.InTypeName = tok.value
    mth.ClientStreaming = true
  } else {
    p.back()
  }
}

sbunce replied:
Refactored both the In and Out stuff to be like this.

dsymonds commented on parser/parser.go:
```diff
@@ -781,15 +781,29 @@ func (p *parser) readService(srv *ast.Service) *parseError {
 		if err := p.readToken(")"); err != nil {
 			return err
 		}
+
```
please avoid unrelated changes; revert this.

sbunce replied:
Done.

Also reverted a few other unrelated changes.

dsymonds commented:
Yes, please just squash and force push. The GH review system is still horrid. I only want to merge a single commit, and this change is small enough to re-review in total each time.

sbunce commented:
I squashed and force pushed.

dsymonds commented on gendesc/gendesc.go:
```diff
@@ -279,6 +279,12 @@ func genMethod(mth *ast.Method) (*pb.MethodDescriptorProto, error) {
 		OutputType: proto.String(qualifiedName(mth.OutType)),
 	}
+	if mth.ClientStreaming {
+		mdp.ClientStreaming = proto.Bool(mth.ClientStreaming)
```
just use `proto.Bool(true)` in these two places.

sbunce replied:
Ah, good point.

Done.

dsymonds commented on parser/parser.go:
```diff
@@ -787,6 +787,21 @@ func (p *parser) readService(srv *ast.Service) *parseError {
+			} else {
+				p.back()
+			}
+		} else {
```
this `else` branch shouldn't be needed now.

sbunce replied:
Thanks! I should have seen that.

dsymonds commented on parser/parser.go:
```diff
@@ -802,6 +817,21 @@ func (p *parser) readService(srv *ast.Service) *parseError {
+			} else {
+				p.back()
+			}
+		} else {
```
ditto

sbunce replied:
Done

dsymonds commented on parser/parser_test.go:
```diff
@@ -335,6 +335,18 @@ var parseTests = []parseTest{
 	},
 	{
+		"SimpleServiceWithMessageCalledStream",
+		"message stream {}\nservice TestService {\n  rpc Foo(stream) returns (stream);\n}",
```
please put the message type at the end of the input, like SimpleService.

sbunce replied:
Done.

sbunce commented:
Thanks for your patience on this review. I should have noticed that extraneous else branch.

I pushed an updated version.

dsymonds commented:
Merged in 6e9e04d.

Thanks for your contribution!
